### PR TITLE
AspectList: Overload __init__ to accept strings

### PR DESCRIPTION
**Pull request by adhikasp**
Overload aspectlist so it could be instanced from a list of string
containing aspects name. The reason of this overloading is to make
it possible to instanced list of desired aspects found in coafile.

```python
>>> aspectlist(['coalaCorrect', 'aspectsYEAH'])
[<aspectclass '...coalaCorrect'>, <aspectclass '...aspectsYEAH'>]
```

Replacement of https://github.com/coala/coala/pull/4379
Closes https://github.com/coala/coala/issues/4382

### CODE REVIEW

File: coalib/bearlib/aspects/collections.py [hunk 1]
```diff
@@ -7,6 +8,21 @@ class aspectlist(list):
...
+        >>> aspectlist([Root.Spelling.coalaCorrect, Root.Spelling.aspectsYEAH])
+        [<aspectclass '...coalaCorrect'>, <aspectclass '...aspectsYEAH'>]
+        >>> aspectlist(['coalaCorrect', 'aspectsYEAH'])
+        [<aspectclass '...coalaCorrect'>, <aspectclass '...aspectsYEAH'>]
+
+        :param seq: A sequence containing either aspectclass, aspectclass
+                    instance, or string of partial/full qualified aspect name.
+        """
+        if all(map((lambda x: isinstance(x, str)), seq)):
+            seq = map((lambda x: coalib.bearlib.aspects[x]), seq)
```
Comment (userzimmermann) [inline coalib/bearlib/aspects/collections.py]: Doesn't cover mixed cases like `aspectlist([Root.Spelling.coalaCorrect, 'aspectsYEAH'])`

Better `super().__init__(item if isaspect(item) else coalib.bearlib.aspects[item] for item in seq)`

File: tests/bearlib/aspects/CollectionsTest.py [hunk 2]
```diff
@@ -9,10 +9,15 @@
 class aspectlistTest:
 
     def test__init__(self):
-        with pytest.raises(aspectTypeError) as exc:
+        list_of_aspect = aspectlist(['CommitMessage.Shortlog',
+                                     'CommitMessage.Body'])
```
Comment (userzimmermann) [inline tests/bearlib/aspects/CollectionsTest.py]: Test aspectclass/instance/string-only and mixed cases

File: coalib/bearlib/aspects/collections.py [hunk 3]
```diff
@@ -7,6 +8,21 @@ class aspectlist(list):
...
     def __init__(self, seq=()):
+        """
+        Initialize a new aspectlist.
+
+        >>> from .root import Root
+        >>> from .Spelling import Spelling
+        >>> aspectlist([Root.Spelling.coalaCorrect, Root.Spelling.aspectsYEAH])
+        [<aspectclass '...coalaCorrect'>, <aspectclass '...aspectsYEAH'>]
+        >>> aspectlist(['coalaCorrect', 'aspectsYEAH'])
+        [<aspectclass '...coalaCorrect'>, <aspectclass '...aspectsYEAH'>]
```
Comment (userzimmermann) [inline coalib/bearlib/aspects/collections.py]: Also add mixed cases

**Comment by adhikasp** [comment]
@userzimmermann updated

File: tests/bearlib/aspects/CollectionsTest.py [hunk 4]
```diff
@@ -9,10 +9,18 @@
...
+        assert isinstance(list_of_aspect, aspectlist)
+        assert list_of_aspect[0] is Metadata.CommitMessage.Shortlog
+        assert list_of_aspect[1] is Metadata.CommitMessage.Body
+        assert list_of_aspect == mix_of_aspect
+
+        with pytest.raises(LookupError) as exc:
             aspectlist(['String'])
-        exc.match("'String' is not an aspectclass or "
-                  'an instance of an aspectclass')
+        exc.match("no aspect named 'String'")
```
Comment (jayvdb) [inline tests/bearlib/aspects/CollectionsTest.py]: exception messages should begin with a capital letter: No aspect named ...

Please raise a bug to fix that.

Comment (jayvdb) [inline tests/bearlib/aspects/CollectionsTest.py]: I've raised https://github.com/coala/coala/issues/4402

File: coalib/bearlib/aspects/collections.py [hunk 5]
```diff
@@ -7,7 +8,23 @@ class aspectlist(list):
     """
 
     def __init__(self, seq=()):
-        super().__init__(map(assert_aspect, seq))
+        """
+        Initialize a new aspectlist.
+
+        >>> from .root import Root
+        >>> from .Spelling import Spelling
+        >>> aspectlist([Root.Spelling.coalaCorrect, Root.Spelling.aspectsYEAH])
```
Comment (jayvdb) [inline coalib/bearlib/aspects/collections.py]: https://github.com/coala/coala/issues/4401 should be fixed first, otherwise lots of unnecessary changes occur.

File: tests/bearlib/aspects/CollectionsTest.py [hunk 6]
```diff
@@ -1,18 +1,28 @@
...
-class AspectListTest:
+class AspectListTest(unittest.TestCase):
 
     def test__init__(self):
-        with pytest.raises(aspectTypeError) as exc:
+        list_of_aspect = AspectList(['CommitMessage.Shortlog',
+                                     'CommitMessage.Body'])
+        mix_of_aspect = AspectList(['CommitMessage.Shortlog',
+                                    Metadata.CommitMessage.Body])
+        assert isinstance(list_of_aspect, AspectList)
```
Comment (jayvdb) [inline tests/bearlib/aspects/CollectionsTest.py]: replace `assert` with `self.assert...(...)` (This was mentioned to you in another PR just merged; fix all of your PRs)

**Comment by jayvdb** [comment]
ack 852ba5e

**Comment by jayvdb** [comment]
@tultor merge

**Comment by jayvdb** [comment]
@rultor merge

**Comment by userzimmermann** [comment]
@adhikasp @jayvdb This was merged a bit too early. It misses tests with aspect instances as mentioned in https://github.com/coala/coala/pull/4389#discussion_r123591451 . But anyway... We can add that later ;)
